### PR TITLE
Fix bugs with WebAuthn preventing sign in and registration.

### LINKED ISSUE #22507: Login with security keys no longer possible
### Description

I get Uncaught DOMException: Failed to execute 'atob' on 'Window': The string to be decoded is not correctly encoded when doing the FIDO2 authentication and cannot login.

Screenshot 2023-01-18 at 7 59 14 AM

### Gitea Version

1.19.0+dev-320-gde484e86b

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

gitea:dev-rootless container

### Database

MySQL

### PR DESCRIPTION
**Pull request by zeripath**
This PR fixes two bugs with Webauthn support:

* There was a longstanding bug within webauthn due to the backend using URLEncodedBase64 but the javascript using decoding using plain base64. This causes intermittent issues with users reporting decoding errors.
* Following the recent upgrade to webauthn there was a change in the way the library expects RPOrigins to be configured. This leads to the Relying Party Origin not being configured and prevents registration.

Fix #22507

Signed-off-by: Andrew Thornton <art27@cantab.net>
Co-authored-by: wxiaoguang <wxiaoguang@gmail.com>

### CODE REVIEW

**Comment by zeripath** [comment]
Alternatively I have raised an issue on https://github.com/WebReflection/uint8-to-base64/issues/1 which may result in an update to the package there and we could use that instead if preferred.

**Review by lunny — APPROVED** [review]

**Comment by zeripath** [comment]
There may be another way of making webpack make this work too.

File: web_src/js/features/user-auth-webauthn.js [hunk 1]
```diff
@@ -1,8 +1,19 @@
 import $ from 'jquery';
-import {encode, decode} from 'uint8-to-base64';
 
 const {appSubUrl, csrfToken} = window.config;
 
+function encodeToBase64(toEncode) {
```
Comment (delvh) [inline web_src/js/features/user-auth-webauthn.js]:
```suggestion
function encodeToBase64(toEncode /* Uint8Array */) {
```

**Review by delvh — APPROVED** [review]

**Comment by wxiaoguang** [comment]
See:

* https://github.com/go-gitea/gitea/issues/22507#issuecomment-1407661615
* #22654

**Review by lunny — CHANGES_REQUESTED** [review]
per @wxiaoguang 's comment

**Comment by lunny** [comment]
It's still error when register a new security key.
```
2023/01/30 09:27:11 ...security/webauthn.go:83:WebauthnRegisterPost() [E] [63d71cef] Unable to finish registration due to error: Error validating origin
	DevInfo: Expected Values: [], Received: https://localhost:3000
2023/01/30 09:27:11 ...security/webauthn.go:85:WebauthnRegisterPost() [E] [63d71cef] CreateCredential: Error validating origin
2023/01/30 09:27:11 ...s/context/context.go:219:HTML() [D] [63d71cef] Template: status/500
2023/01/30 09:27:11 ...user/notification.go:45:func1() [I] [63d71cef] [SQL] SELECT count(*) FROM "notification" WHERE (user_id = $1) AND (status = $2) [3 1] - 12.352889ms
2023/01/30 09:27:11 [63d71cef] router: completed POST /user/settings/security/webauthn/register for [::1]:55056, 500 Internal Server Error in 45.6ms @ security/webauthn.go:62(security.WebauthnRegisterPost)
```

**Comment by wxiaoguang** [comment]
Yup, that's why I said in my demo PR: "After: The dialog can be displayed.", but not "It works". Because it still doesn't work ....

I just doubt that whether any one has really tested it.

I am glad to see the error is the same now: "Error validating origin" (my is on `https://production`, when login with an existing webauthn)

**Comment by lunny** [comment]
> Yup, that's why I said in my demo PR: "After: The dialog can be displayed.", but not "It works". Because it still doesn't work ....
> 
> I just doubt that whether any one has really tested it.
> 
> I am glad to see the error is the same now: "Error validating origin" (my is on `https://production`, when login with an existing webauthn)

Looks like, even revert #22400 , the error is still there.

**Comment by wxiaoguang** [comment]
You also need this to make webauthn (new lib) work (the reason is complicated, so keep this comment simple)

I have tested on my side, I can login with webauthn.

```diff
diff --git a/modules/auth/webauthn/webauthn.go b/modules/auth/webauthn/webauthn.go
index d08f7bf7c..937da872c 100644
--- a/modules/auth/webauthn/webauthn.go
+++ b/modules/auth/webauthn/webauthn.go
@@ -28,7 +28,7 @@ func Init() {
                Config: &webauthn.Config{
                        RPDisplayName: setting.AppName,
                        RPID:          setting.Domain,
-                       RPOrigin:      appURL,
+                       RPOrigins:     []string{appURL},
                        AuthenticatorSelection: protocol.AuthenticatorSelection{
                                UserVerification: "discouraged",
                        },
```

File: web_src/js/features/user-auth-webauthn.js [hunk 2]
```diff
@@ -85,14 +96,21 @@ function verifyAssertion(assertedCredential) {
   });
 }
 
-// Encode an ArrayBuffer into a base64 string.
-function bufferEncode(value) {
-  return encode(value)
+// Encode an ArrayBuffer into a URLEncoded base64 string.
+function bufferURLEncodedBase64(value) {
```
Comment (wxiaoguang) [inline web_src/js/features/user-auth-webauthn.js]: Maybe it could be `encodeURLEncodedBase64` like `decodeURLEncodedBase64`

**Comment by zeripath** [comment]
> I just doubt that whether any one has really tested it.

This definitely used to work and was tested thoroughly.

**Comment by zeripath** [comment]
Thanks for the RPOrigins thing - that explains the sudden breakage!

File: web_src/js/features/user-auth-webauthn.js (hunk 2 above)
Comment (zeripath) [reply web_src/js/features/user-auth-webauthn.js]: done

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
> git checkout -b zeripath-fix-atob-on-window main

CI failure is related.

**Comment by james-d-elliott** [comment]
Let me know if I can assist in any way with this, regarding the javascript I'd recommend @simplewebauthn/browser.

**Review by lunny — APPROVED** [review]

**Comment by zeripath** [comment]
Make lgtm work

**Comment by yardenshoham** [comment]
Can't really backport without #22400

**Comment by yardenshoham** [comment]
Do these bugs exist in v1.18? Maybe we don't have to backport

**Comment by delvh** [comment]
They shouldn't exist on 1.18.
As I said on the issue, I can still log in without a problem to a private 1.18 instance.
The problem appears to be mainly due to the upgraded library, and the illegal behavior was probably ignored previously.

**Comment by delvh** [comment]
The only thing I am somewhat afraid is that logins that previously valid suddenly stop being valid…
After all, the output sent to the server is now different?

**Comment by zeripath** [comment]
Nah, the output sent is the same - it's the response received by the js that is different and those that are different would have previously caused spurious errors.

This PR does fix some another issue that was not related to the upgraded library so there will need to be a partial backport primarily of the JS changes only (skipping the RPOrigins stuff). I should be able to get that done tonight.

**Pull request merged**